### ISSUE TITLE
"unknown rank" error" in tensorflow.keras.layers.Layer with tensorflow.py_function output

### ISSUE THREAD

**Issue by MegaCreater** [tensorflow/tensorflow#59726]
### Issue Type

Bug

### Have you reproduced the bug with TF nightly?

No

### Source

source

### Tensorflow Version

v2.11.0-rc2-17-gd5b57ca93e5 2.11.0

### Custom Code

Yes

### OS Platform and Distribution

Google colab 

### Current Behaviour?

```shell
A bug happened!
```

### Standalone code to reproduce the issue

```shell
Ss
```

**Comment by pjpratik** [comment]
I was able to reproduce this issue in TF 2.11 and TF Nightly 2.13.0-dev20230219 . Please find the gist [here](https://colab.research.google.com/gist/pjpratik/20cd82f24b7697da535ba222cd066b28/59726.ipynb). 

@SuryanarayanaY Could you please look into this issue. Thanks.

**Comment by SuryanarayanaY** [comment]
Hi @MegaCreater ,
The error might be due to python function using inside the Graph for which the shape is unknown.Hence getting the error:Unknown rank error.
The proposed workaround is you have add this code `inputs.set_shape([None,32,32,3])` after tf.nupmy_function like below.
@tf.function
```
def sample(inputs:tf.Tensor):                                                   # some sample function 

  def numpy_method(inputs:np.ndarray):
    # some complex method 
    return np.sqrt(inputs)# example code 

  inputs=tf.numpy_function(numpy_method,[inputs],inputs.dtype,name='outputs')
  inputs.set_shape([None,32,32,3])                                #New line of code
  inputs=tf.keras.layers.AvgPool2D()(inputs)                                   

  return inputs

sample(tf.random.uniform((1,32,32,3)))
```

Please refer to the attached gist [here](https://colab.research.google.com/gist/SuryanarayanaY/5c64d66b14b62ca6308674e4b80f6a13/59726.ipynb).

Hope this will solve your purpose. Thanks!

**Comment by MegaCreater** [comment]
Can we update something here -> https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/ops/script_ops.py#L768 
somthing like this -> 

```python
# ...... 
def numpy_function(func, inp, Tout, outputs_shapes:tuple=None, stateful=True, name=None):

  ''' docs ------ 
  '''

  outputs = py_func_common(func, inp, Tout, stateful=stateful, name=name)# line 768
  if isinstance(outputs,tf.Tensor) and (outputs.shape==None):
    if outputs_shapes==None: 
      if isinstance(inp,(list,tuple)): outputs.set_shape(inp[0].shape)
      else: outputs.set_shape(inp.shape)
    else: outputs.set_shape(outputs_shapes)
  #elif isinstance(outputs,(list,tuple)): ..... 
  return outputs

# ..... 
```

**Comment by SuryanarayanaY** [comment]
Hi @MegaCreater ,

I have tested the proposed changes and seems working fine for this case.Please refer to attached tested [gist](https://colab.research.google.com/gist/SuryanarayanaY/f032311e9975716a69fc06d8ba0c06a0/59726_r3-code-gtest.ipynb). At least the changes works fine with and without @tf.function. I am not sure whether this covers all cases or not. Also it involves the structural changes in API as it adds new argument also.This may needs to be reviewed properly for short falls if any.

The proposed changes relies completely on input_shape which we are assuming that output_shape also being same as input_shape. What about the case when there is change in shapes that actually happens inside the numpy_method ?

Could we cover this to all use cases ?  Thanks!

**Comment by MegaCreater** [comment]
Yes we can. When the shapes of output is same then user wont have to pass `output_shape` variable (i.e `output_shape=None`, but when `output_shape` changes then `output_shape` has to be provided. Give me some time to work on it more. I will update you. 
But still with this update `rank` issue is not resolved. I am look for that also.

**Comment by SuryanarayanaY** [comment]
@MegaCreater ,

If `output_shape` shape has to be passed by the user explicitly when output_shape changes based on computation then the same thing can be done by `inputs.set_shape([None,32,32,3]) ` that proposed in earlier workaround right.

I would like to understand how it is different from earlier workaround. For me the earlier work around using `set_shape` might work for all cases.Is there any exception for this ?

**Comment by Anan5a** [comment]
I feel like this should be documented officially. I just wasted a few hours pulling my hairs :(
